File: commands/hf.js
"use strict";
/**
 * commands/hf.js  —  HuggingFace Multi-Model AI
 * ──────────────────────────────────────────────────
 * REFACTORED: حُذف sessionSchema/loadCtx/saveCtx/downloadAsBase64/detectAttachment
 *             (~110 سطراً) → utils/aiSession.js + utils/mediaUtils.js
 */

const axios  = require("axios");
const { loadSession, saveSession, clearSession } = require("../utils/aiSession");
const { detectAttachment, downloadImageAsBase64, getSenderName } = require("../utils/mediaUtils");

const HF_BASE  = (process.env.HF_SPACE_URL || "").replace(/\/+$/, "");
const NS       = "hf";
const MAX_MSGS = 10;

// ─── تنظيف وسوم التفكير من الرد ─────────────────────────────
function cleanReply(text) {
  text = text.replace(/<(?:think|thinking|analysis|reflection)>[\s\S]*?<\/(?:think|thinking|analysis|reflection)>/gi, "");
  const match = text.match(/(?:الجواب|الإجابة|Answer)\s*:\s*/i);
  if (match) text = text.slice(text.indexOf(match[0]) + match[0].length);
  return text.trim();
}

async function callHF(messages, model) {
  if (!HF_BASE) throw new Error("HF_SPACE_URL غير مضبوط في متغيرات البيئة");
  const { data } = await axios.post(
    `${HF_BASE}/hf`,
    { messages, model, max_tokens: 512 },
    { timeout: 65000, headers: { "Content-Type": "application/json" } }
  );
  if (data.error) throw new Error(data.error);
  if (!data.reply) throw new Error("استجابة فارغة من الخادم");
  return { reply: cleanReply(data.reply), model_used: data.model_used || model };
}

async function handle(api, event, args, registerReply) {
  const { threadID, messageID, senderID } = event;
  const firstArg = args[0]?.toLowerCase() || "";

  // ─── مسح الذاكرة ────────────────────────────────────────────
  if (["مسح", "clear", "reset"].includes(firstArg)) {
    await clearSession(NS, threadID);
    return api.sendMessage("🧹 تم مسح ذاكرة المجموعة.", threadID, null, messageID);
  }

  // ─── عرض النماذج ─────────────────────────────────────────────
  if (["نماذج", "models", "list"].includes(firstArg)) {
    return api.sendMessage(
      `🤖 النماذج المتاحة في HF AI:\n\n` +
      `━━━━━ Llama (Meta) ━━━━━\n• llama4 ← الافتراضي ✅ يدعم الصور\n• llama / llama8 → Llama-3.1-8B\n• llama70 → Llama-3.3-70B\n\n` +
      `━━━━━ Qwen (Alibaba) ━━━━━\n• qwen7 → Qwen2.5-7B\n• qwen / qwen72 → Qwen2.5-72B\n• qwen3 → Qwen3-235B\n\n` +
      `━━━━━ Mistral ━━━━━\n• mistral → Mistral-7B\n• mistral22 → Mistral-Small-22B ✅ يدعم الصور\n• mixtral → Mixtral-8x7B\n\n` +
      `━━━━━ Google ━━━━━\n• gemma → Gemma-3-27B ✅ يدعم الصور\n• gemma4 → Gemma-3-4B ✅ يدعم الصور\n\n` +
      `━━━━━ DeepSeek ━━━━━\n• deepseek7 → DeepSeek-R1-7B\n• deepseek → DeepSeek-R1-32B\n\n` +
      `━━━━━ أخرى ━━━━━\n• phi / phi4 → Microsoft Phi\n• zephyr → Zephyr-7B\n• command → Cohere Command-R\n\n` +
      `📸 النماذج التي تدعم الصور: llama4، mistral22، gemma، gemma4\n` +
      `🧹 .hf مسح — مسح الذاكرة`,
      threadID, null, messageID
    );
  }

  // ─── تحديد النموذج والسؤال ───────────────────────────────────
  const { messages: savedCtx, model: savedModel } = await loadSession(NS, threadID, MAX_MSGS);

  const looksLikeModel = firstArg && !firstArg.includes(" ") &&
    !/^[\u0600-\u06FF]/.test(firstArg) &&
    (firstArg.includes("/") || firstArg.length <= 15);

  let model, promptArgs;
  if (looksLikeModel && args.length >= 1) {
    model      = args[0];
    promptArgs = args.slice(1);
  } else {
    model      = savedModel || "llama4";
    promptArgs = args;
  }

  let prompt = promptArgs.join(" ").trim();
  if (!prompt && event.messageReply?.body) prompt = event.messageReply.body.trim();

  const attachment = detectAttachment(event);

  if (!prompt && !attachment) {
    return api.sendMessage(
      `🤖 HF AI — النموذج الحالي: ${model}\n\n` +
      `.hf [سؤالك]  ← ${model} افتراضي\n` +
      `.hf qwen7 اشرح البرمجة\n` +
      `.hf gemma + صورة — تحليل الصورة\n` +
      `.hf نماذج — كل النماذج\n` +
      `.hf مسح — مسح الذاكرة`,
      threadID, null, messageID
    );
  }

  if (!HF_BASE)
    return api.sendMessage("❌ HF_SPACE_URL غير مضبوط في متغيرات البيئة", threadID, null, messageID);

  // ─── رسالة حالة + اسم مرسل (بالتوازي) ──────────────────────
  const [sentMsg, senderName] = await Promise.all([
    new Promise(res => api.sendMessage(
      attachment
        ? `⏳ جاري تحليل ${attachment.kind === "image" ? "الصورة 🖼️" : attachment.kind === "audio" ? "الصوت 🎵" : "الفيديو 🎬"} لـ ${model}...`
        : `⏳ جاري السؤال لـ ${model}...`,
      threadID, (e, i) => res(i), messageID
    )),
    getSenderName(api, senderID),
  ]);

  const statusMsgId = sentMsg?.messageID;
  const updateStatus = async (text) => {
    try { if (statusMsgId) await api.editMessage(text, statusMsgId); } catch (_) {}
  };

  let userMsg;
  if (attachment?.kind === "image") {
    const imgData = await downloadImageAsBase64(attachment.url).catch(() => null);
    userMsg = imgData
      ? { role: "user", content: `[${senderName}]: ${prompt || "صف هذه الصورة"}`, attachment: { kind: "image", base64: imgData.data, contentType: imgData.mediaType } }
      : { role: "user", content: `[${senderName}]: ${prompt || "صف هذه الصورة"}` };
    if (!imgData) await updateStatus("⚠️ تعذّر تحميل الصورة، سأجيب على النص فقط...");
  } else if (attachment) {
    userMsg = { role: "user", content: `[${senderName}]: ${prompt || (attachment.kind === "audio" ? "فرّغ هذا الصوت" : "حلل هذا الفيديو")}`, attachment: { kind: attachment.kind, url: attachment.url } };
  } else {
    userMsg = { role: "user", content: `[${senderName}]: ${prompt}` };
  }

  try {
    const { reply, model_used } = await callHF([...savedCtx, userMsg], model);

    await updateStatus(reply);

    if (statusMsgId && registerReply) {
      registerReply(statusMsgId, { author: senderID }, async ({ api, event }) => {
        await handle(api, event, [model, event.body?.trim() || ""].filter(Boolean), registerReply);
      });
    }

    await saveSession(NS, threadID, [
      ...savedCtx,
      { role: "user",      content: userMsg.content },
      { role: "assistant", content: reply },
    ], model_used, MAX_MSGS);

  } catch (err) {
    let msg = "❌ خطأ: ";
    if (err.code === "ECONNABORTED" || err.message?.includes("timeout"))
      msg += "⏱️ انتهت مهلة الاتصال — جرب نموذجاً أصغر مثل: llama أو gemma4";
    else
      msg += (err.message || "فشل الاتصال").substring(0, 120);
    await updateStatus(msg);
  }
}

module.exports = {
  config: {
    name:             "hf",
    aliases:          ["huggingface", "hfai"],
    version:          "3.2.0",
    author:           "Sunken",
    countDown:        5,
    role:             0,
    shortDescription: { ar: "ذكاء اصطناعي — llama4 افتراضي + دعم الصور" },
    longDescription:  { ar: "تحدث مع أي نموذج من HuggingFace\nالنموذج الافتراضي: llama4 (يدعم الصور ✅)" },
    category:         "ذكاء اصطناعي",
    guide:            { ar: "{pn}hf [سؤالك]\n{pn}hf qwen7 اشرح البرمجة\n{pn}hf نماذج\n{pn}hf مسح" },
  },

  onStart: async ({ api, event, args, message }) => {
    await handle(api, event, args, message?.registerReply);
  },

  onReply: async ({ api, event, message }) => {
    await handle(api, event, [event.body?.trim() || ""], message?.registerReply);
  },
};

File: plugins/hf.py
"""
plugins/hf.py
endpoint: POST /hf
يدعم النصوص + الصور (base64 vision)
"""

import os
import logging
import re
from huggingface_hub import InferenceClient
from fastapi import Request
from fastapi.responses import JSONResponse

logger = logging.getLogger(__name__)

DESCRIPTION = "HuggingFace Inference API — نصوص + صور"

HF_TOKEN = os.environ.get("HF_TOKEN", "")

SHORTCUTS: dict[str, str] = {
    "qwen":      "Qwen/Qwen2.5-72B-Instruct",
    "qwen72":    "Qwen/Qwen2.5-72B-Instruct",
    "qwen7":     "Qwen/Qwen2.5-7B-Instruct",
    "qwen3":     "Qwen/Qwen3-235B-A22B",
    "llama":     "meta-llama/Llama-3.1-8B-Instruct",
    "llama70":   "meta-llama/Llama-3.3-70B-Instruct",
    "llama8":    "meta-llama/Llama-3.1-8B-Instruct",
    "llama4":    "meta-llama/Llama-4-Scout-17B-16E-Instruct",
    "mistral":   "mistralai/Mistral-7B-Instruct-v0.3",
    "mistral22": "mistralai/Mistral-Small-3.1-22B-Instruct-2503",
    "mixtral":   "mistralai/Mixtral-8x7B-Instruct-v0.1",
    "deepseek":  "deepseek-ai/DeepSeek-R1-Distill-Qwen-32B",
    "deepseek7": "deepseek-ai/DeepSeek-R1-Distill-Qwen-7B",
    "phi":       "microsoft/Phi-3.5-mini-instruct",
    "phi4":      "microsoft/phi-4",
    "gemma":     "google/gemma-3-27b-it",
    "gemma4":    "google/gemma-3-4b-it",
    "zephyr":    "HuggingFaceH4/zephyr-7b-beta",
    "command":   "CohereForAI/c4ai-command-r-plus-08-2024",
}

VISION_FALLBACK = "meta-llama/Llama-4-Scout-17B-16E-Instruct"

VISION_MODELS = {
    "meta-llama/Llama-4-Scout-17B-16E-Instruct",
    "google/gemma-3-27b-it",
    "google/gemma-3-4b-it",
    "mistralai/Mistral-Small-3.1-22B-Instruct-2503",
}

SYSTEM_PROMPT = (
    'أنت بوت مساعد ذكي اسمك "Sunken". '
    'أجب دائماً باللغة العربية بإيجاز (أقل من 300 كلمة). '
    'كن ودوداً ومهذباً. '
    'لا تكتب أي تفكير أو تحليل داخلي، اكتب الجواب النهائي فقط مباشرة.'
)


def resolve_model(name: str) -> str:
    key = name.lower().strip()
    if key in SHORTCUTS:
        return SHORTCUTS[key]
    if "/" in name:
        return name
    for k, v in SHORTCUTS.items():
        if k.startswith(key) or key in k:
            return v
    return name


def clean_reply(text: str) -> str:
    """
    إزالة كل أنماط التفكير والتحليل الداخلي:
    - <think>...</think>
    - <thinking>...</thinking>
    - <analysis>...</analysis>
    - أي نص قبل الجواب الحقيقي يبدأ بـ "الجواب:" أو "Answer:" إذا وُجد
    """
    # حذف وسوم التفكير
    text = re.sub(r"<think>[\s\S]*?</think>", "", text, flags=re.IGNORECASE)
    text = re.sub(r"<thinking>[\s\S]*?</thinking>", "", text, flags=re.IGNORECASE)
    text = re.sub(r"<analysis>[\s\S]*?</analysis>", "", text, flags=re.IGNORECASE)
    text = re.sub(r"<reflection>[\s\S]*?</reflection>", "", text, flags=re.IGNORECASE)

    # إذا فيه "الجواب:" أو "Answer:" → خذ ما بعده فقط
    match = re.search(r"(?:الجواب|الإجابة|Answer)\s*:\s*", text, flags=re.IGNORECASE)
    if match:
        text = text[match.end():]

    return text.strip()


def has_image(messages: list) -> bool:
    return any(
        m.get("attachment", {}).get("kind") == "image" and m.get("attachment", {}).get("base64")
        for m in messages
    )


def build_messages(raw_messages: list, model_id: str) -> list:
    result = []

    if not any(m.get("role") == "system" for m in raw_messages):
        result.append({"role": "system", "content": SYSTEM_PROMPT})

    for msg in raw_messages:
        role = msg.get("role", "user")
        text = msg.get("content", "")
        att  = msg.get("attachment")

        if att and att.get("kind") == "image" and att.get("base64"):
            if model_id in VISION_MODELS:
                content = [
                    {
                        "type": "image_url",
                        "image_url": {
                            "url": f"data:{att['contentType']};base64,{att['base64']}"
                        }
                    },
                    {"type": "text", "text": text or "وصف هذه الصورة"},
                ]
            else:
                content = f"[المستخدم أرسل صورة] {text or 'وصف هذه الصورة'}"
        else:
            content = text

        result.append({"role": role, "content": content})

    return result


def call_hf_sync(model_id: str, messages: list, max_tokens: int = 512) -> tuple[str, str]:
    token = HF_TOKEN.strip()
    if not token:
        raise RuntimeError("HF_TOKEN غير موجود — أضفه في Settings → Variables and secrets")

    actual_model = model_id
    if has_image(messages) and model_id not in VISION_MODELS:
        logger.warning(f"[hf] {model_id} لا يدعم Vision → تحويل لـ {VISION_FALLBACK}")
        actual_model = VISION_FALLBACK

    client = InferenceClient(model=actual_model, token=token)
    built  = build_messages(messages, actual_model)

    result = client.chat_completion(
        messages=built,
        max_tokens=max_tokens,
        temperature=0.7,
    )

    reply = clean_reply(result.choices[0].message.content or "")
    if not reply:
        raise RuntimeError("استجابة فارغة من النموذج")
    return reply, actual_model


def register(app):

    @app.post("/hf")
    async def hf_endpoint(request: Request):
        import asyncio
        try:
            body       = await request.json()
            model_raw  = body.get("model", "llama4")
            messages   = body.get("messages", [])
            max_tokens = int(body.get("max_tokens", 512))

            if not messages:
                return JSONResponse({"error": "messages مطلوب"}, status_code=400)

            model_id = resolve_model(model_raw)
            logger.info(f"[hf] {model_raw} → {model_id} | has_image={has_image(messages)}")

            try:
                loop = asyncio.get_event_loop()
                reply, model_used = await loop.run_in_executor(
                    None, call_hf_sync, model_id, messages, max_tokens
                )
                return JSONResponse({"reply": reply, "model_used": model_used})

            except Exception as e:
                logger.error(f"[hf] error: {e}")
                return JSONResponse({"error": str(e), "model_used": model_id}, status_code=503)

        except Exception as e:
            logger.exception(f"[hf] Exception: {e}")
            return JSONResponse({"error": str(e)[:200]}, status_code=500)

    @app.get("/hf/models")
    async def hf_models():
        return JSONResponse({
            "shortcuts": SHORTCUTS,
            "vision_models": list(VISION_MODELS),
            "vision_fallback": VISION_FALLBACK,
            "default_model": "llama4",
        })
